File: src/main/java/com/checkers/board/Board.java
package com.checkers.board;

import com.checkers.exceptions.CannotEatException;
import com.checkers.exceptions.CannotMoveException;
import com.checkers.exceptions.CellIsBusyException;
import com.checkers.exceptions.CellIsEmptyException;
import com.checkers.user.User;
import com.checkers.utils.Color;
import lombok.Getter;

public class Board {

    private static final int SIZE_OF_BOARD = 8;

    @Getter
    private final Cell[][] board = new Cell[SIZE_OF_BOARD][SIZE_OF_BOARD];

    private int whitePieces = 12;

    private int blackPieces = 12;

    public Board() {
        createCellsInBoard();
        initWhitePieces();
        initBlackPieces();
        analyzeAbilities();
    }

    private void createCellsInBoard() {
        for (int row = 0; row < SIZE_OF_BOARD; row++) {
            for (int col = 0; col < SIZE_OF_BOARD; col++) {
                board[row][col] = new Cell(row + 1, col + 1, this);
            }
        }
    }

    private void initWhitePieces() {
        for (int row = 0; row < SIZE_OF_BOARD / 2 - 1; row++) {
            for (int col = 0; col < SIZE_OF_BOARD; col++) {
                if (board[row][col].getColor() == Color.BLACK) {
                    Piece newPiece = new Man(Color.WHITE);
                    board[row][col].setPiece(newPiece);
                    newPiece.setCell(board[row][col]);
                }
            }
        }
    }

    private void initBlackPieces() {
        for (int row = SIZE_OF_BOARD - 1; row > SIZE_OF_BOARD / 2; row--) {
            for (int col = 0; col < SIZE_OF_BOARD; col++) {
                if (board[row][col].getColor() == Color.BLACK) {
                    Piece newPiece = new Man(Color.BLACK);
                    board[row][col].setPiece(newPiece);
                    newPiece.setCell(board[row][col]);
                }
            }
        }
    }

    private void analyzeAbilities() {
        for (int row = 0; row < SIZE_OF_BOARD; row++) {
            for (int col = 0; col < SIZE_OF_BOARD; col++) {
                Piece piece = board[row][col].getPiece();
                if (piece != null) {
                    piece.analyzeAbilityOfMove();
                    piece.analyzeAbilityOfEat();
                }
            }
        }
    }

    public Cell getCell(int row, int col) {
        if (row < 1 || row > 8 || col < 1 || col > 8) {
            return null;
        }

        return board[row - 1][col - 1];
    }

    public void move(Cell sourceCell, Cell destinationCell) {
        if (sourceCell.isEmpty()) {
            throw new CellIsEmptyException(sourceCell);
        }

        if (!destinationCell.isEmpty()) {
            throw new CellIsBusyException(destinationCell);
        }

        Piece piece = sourceCell.getPiece();
        if (piece.isAbleToMoveTo(destinationCell)) {
            piece.move(destinationCell);
        } else {
            throw new CannotMoveException(sourceCell, destinationCell);
        }
    }

    public void eat(Cell sourceCell, Cell destinationCell) {
        if (sourceCell.isEmpty()) {
            throw new CellIsEmptyException(sourceCell);
        }

        if (!destinationCell.isEmpty()) {
            throw new CellIsBusyException(destinationCell);
        }

        var piece = sourceCell.getPiece();
        if (piece.isAbleToEatTo(destinationCell)) {
            piece.eat(destinationCell);
        } else {
            throw new CannotEatException(sourceCell, destinationCell);
        }
    }

    public void analyze(User user) {
        Color userColor = user.getColor();

        for (int row = 0; row < SIZE_OF_BOARD; row++) {
            for (int col = 0; col < SIZE_OF_BOARD; col++) {
                if (board[row][col].getPiece() != null) {
                    board[row][col].getPiece().analyzeAbilityOfMove();
                    board[row][col].getPiece().analyzeAbilityOfEat();
                    if (!board[row][col].getPiece().isCanMove()) {
                        System.out.println(row + " " + col);
                    }
                }
            }
        }

        boolean isCanEat = false;
        for (int row = 1; row <= SIZE_OF_BOARD && !isCanEat; row++) {
            for (int col = 1; col <= SIZE_OF_BOARD && !isCanEat; col++) {
                Cell cell = getCell(row, col);
                if (cell.getColor() == Color.WHITE) continue;
                Piece userPiece = cell.getPiece();
                if (userPiece == null) continue;
                if (userPiece.getColor() == userColor && userPiece.isCanEat()) {
                    isCanEat = true;
                }
            }
        }
        user.setCanEat(isCanEat);
    }

    public boolean isGaming() {
        return whitePieces != 0 && blackPieces != 0;
    }
}

File: src/main/kotlin/com/checkers/board/Board.kt
package com.checkers.board

import com.checkers.exception.*
import com.checkers.user.User
import com.checkers.utils.BoardUtils
import com.checkers.utils.Color

class Board {

    val board: Array<Array<Cell>> = Array(SIZE_OF_BOARD) { row ->
        Array(SIZE_OF_BOARD) { column ->
            Cell(row + 1, column + 1, this)
        }
    }

    private var whitePieces = 12

    private var blackPieces = 12

    init {
        initWhitePieces()
        initBlackPieces()
        analyzeAbilities()
    }

    private fun initWhitePieces() =
        (0..2).asSequence()
            .flatMap { board[it].asSequence() }
            .filter { it.color == Color.BLACK }
            .forEach { it.piece = Man(Color.WHITE) }

    private fun initBlackPieces() =
        (5..7).asSequence()
            .flatMap { board[it].asSequence() }
            .filter { it.color == Color.BLACK }
            .forEach { it.piece = Man(Color.BLACK) }

    private fun cells(): Sequence<Cell> =
        board.asSequence()
            .flatMap { it.asSequence() }

    private fun pieces(): Sequence<Piece> =
        cells()
            .mapNotNull { it.piece }

    fun getCell(row: Int, column: Int): Cell =
        if (BoardUtils.isCoordinatesExists(row, column)) {
            board[row - 1][column - 1]
        } else {
            throw CellNotFoundException(row, column)
        }

    fun move(sourceCell: Cell, destinationCell: Cell) {
        val piece = sourceCell.piece
            ?: throw CellIsEmptyException(sourceCell)

        if (destinationCell.piece != null) {
            throw CellIsBusyException(destinationCell)
        }

        if (piece.isAbleToMoveTo(destinationCell)) {
            piece.move(destinationCell)
        } else {
            throw CannotMoveException(sourceCell, destinationCell)
        }
    }

    fun eat(sourceCell: Cell, destinationCell: Cell) {
        val piece = sourceCell.piece
            ?: throw CellIsEmptyException(sourceCell)

        if (destinationCell.piece != null) {
            throw CellIsBusyException(destinationCell)
        }

        if (piece.isAbleToEatTo(destinationCell)) {
            piece.eat(destinationCell)
        } else {
            throw CannotEatException(sourceCell, destinationCell)
        }
    }

    fun analyze(user: User) {
        val userColor = user.color

        analyzeAbilities()

        user.isCanEat = pieces()
            .any { it.color == userColor && it.isCanEat }
    }

    private fun analyzeAbilities() =
        pieces()
            .forEach {
                it.analyzeAbilityOfMove()
                it.analyzeAbilityOfEat()
            }

    fun isGaming(): Boolean =
        whitePieces != 0 && blackPieces != 0

    fun decrementWhitePieceCount() {
        if (whitePieces > 0) {
            whitePieces--
        }
    }

    fun decrementBlackPieceCount() {
        if (blackPieces > 0) {
            blackPieces--
        }
    }

    companion object {
        private const val SIZE_OF_BOARD = 8
    }
}